app_dart/lib/src/model/firestore/ci_staging.dart

```diff
@@ -0,0 +1,157 @@
+  /// signals how many more tests are running. Returns the record (valid: false)
+  /// otherwise.
+  static Future<StagingConclusion> markConclusion({
+    required FirestoreService firestoreService,
+    // required String documentNameString,
+    required RepositorySlug slug,
+    required String sha,
+    required String stage,
+    required String checkRun,
+    required String conclusion,
+  }) async {
+    final logCrumb = 'markConclusion(${slug.owner}_${slug.name}_${sha}_$stage, $checkRun, $conclusion)';
+
+    // Marking needs to happen while in a transaction to ensure `remaining` is
+    // updated correctly. For that to happen correctly; we need to as perform a
+    // read of the document in the transaction as well. So start the transaction
+    // first thing.
+    final docRes = await firestoreService.documentResource();
+    final transactionResponse = await docRes.beginTransaction(
+      BeginTransactionRequest(
+        options: TransactionOptions(
+          readWrite: ReadWrite(),
+        ),
+      ),
+      kDatabase,
+    );
+    final transaction = transactionResponse.transaction;
+    if (transaction == null) {
+      log.warning('$logCrumb: transaction was null');
+      throw '$logCrumb: transaction was null';
+    }
+
+    var newRemaining = -1;
+
+    late final Document doc;
+
+    // transaction block
+    try {
+      // First: read the fields we want to change.
+      final documentPath = documentNameFor(sha: sha, slug: slug, stage: stage);
+      doc = await docRes.get(documentPath, mask_fieldPaths: [kRemainingField, checkRun], transaction: transaction);
+
+      // Fields and remaining _must_ be present.
+      if (doc.fields == null ||
+          doc.fields![kRemainingField] == null ||
+          doc.fields![kRemainingField]!.integerValue == null) {
+        log.warning('$logCrumb: missing field for $transaction / ${doc.fields}');
+        throw '$logCrumb missing fields';
+      }
+
+      final remaining = int.parse(doc.fields![kRemainingField]!.integerValue!);
+      newRemaining = remaining - 1;
+      // We will have check_runs scheduled after the engine was built successfully, so missing the checkRun field
+      // is an OK response to have. All fields should have been written at creation time.
+      if (doc.fields?[checkRun] == null || doc.fields?[checkRun]!.stringValue == null) {
+        log.info('$logCrumb: $checkRun not present in doc; remaining=$remaining');
+        await docRes.rollback(RollbackRequest(transaction: transaction), kDatabase);
+        return (valid: false, remaining: remaining);
+      }
+
+      // Now we can modify the document the change in the conculsion
+      if (kDefaultTaskStatus == doc.fields![checkRun]!.stringValue) {
+        log.info('$logCrumb: setting remaining to $newRemaining and changing ${doc.fields![checkRun]!.stringValue}');
+        doc.fields![checkRun] = Value(stringValue: conclusion);
```

Comment:
A little bit uneasy about storing check run as just conclusion string. It will be difficult to add more data later. Maybe wrap it in a field? Something like:

```dart
doc.fields![checkRun]!['conclusion'] = Value(stringValue: conclusion);
```

Then in the future it would trivial to add something like:

```dart
doc.fields![checkRun]!['finishDateTime'] = ...;
doc.fields![checkRun]!['isRequired'] = ...;
...
```

Reply:
What you're asking for is a map - I'd have to check to see how firestore actually send that down as data - and also how the filters work with that.

Reply:
Map value issue: 

```
/// The map keys, represented as UTF-8,
/// must not exceed 1,500 bytes and cannot be empty.
```
  
Assuming that comment is correct for MapValue (firebase maps!), I'm uncomfortable with adding this as a landmine for the future (I added a test and it exploded).  I view staging as "must build" already; but we could have an array later?

Reply:
Marking as resolved for now. We can make changes in the future; but staging fields are "required" right now.  I'm not sure how / why we'd want non-required fields, but we can make a new document to encode that in the future.